### PR TITLE
Adding functional tests for Custom-Instance-Attributes

### PR DESCRIPTION
**Pull request by vsiddharth**
* This patch tests the happy paths of the custom-instance-attribute
 feature. The `ECS_INSTANCE_ATTRIBUTE` holds a set of key=value pairs
 which are exposed via the DescribeContainerInstances API. Each key or value
 can contain upto 128 characters and we currently only support a maximum
 of 10 custom attributes per instance.

http://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_PutAttributes.html
describes more details about the limits for the attributes.

Signed-off-by: Vinothkumar Siddharth <sidvin@amazon.com>

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 

### Licensing

This contribution is under the terms of the Apache 2.0 License:

### CODE REVIEW

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 1]
```diff
@@ -618,3 +618,197 @@ func TestNetworkModeBridge(t *testing.T) {
...
 	require.NoError(t, err, "Networking mode bridge testing failed")
 }
+
+// TestCustomAttributesWithValidOptions tests the ECS_INSTANCE_ATTRIBUTES
+// upon agent registration with a key, value pair
+func TestCustomAttributesWithValidOptions(t *testing.T) {
+	len := 25
+	key, value := GetRandomString(len), GetRandomString(len)
+
+	customAttribute := fmt.Sprintf("{\"%s\": \"%s\"}", key, value)
```
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: [multiple occurrences] Why not use a const for this? What are we gaining by generating a random key, val for this? We'd have significantly less loc's, which would also mean less potential for bugs.

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 2]
```diff
@@ -618,3 +618,197 @@ func TestNetworkModeBridge(t *testing.T) {
...
+	os.Setenv("ECS_INSTANCE_ATTRIBUTES", customAttribute)
```
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: [multiple occurrences] can you add a `defer os.UnsetEnv("ECS_INSTANCE_ATTRIBUTES")` here?

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 3]
```diff
@@ -618,3 +618,197 @@ func TestNetworkModeBridge(t *testing.T) {
...
+		Cluster:            &agent.Cluster,
+		ContainerInstances: []*string{&agent.ContainerInstanceArn},
+	}
+
+	resp, err := ECS.DescribeContainerInstances(params)
+	require.NoError(t, err)
+	require.NotEmpty(t, resp.ContainerInstances)
+
+	attribMap := AttributesToMap(resp.ContainerInstances[0].Attributes)
+	assert.Equal(t, value, attribMap[key], "Values should match")
```
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: [multiple occurrences] can you do an `assert.NotEmpty(attribMap)` here?

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 4]
```diff
@@ -618,3 +618,197 @@ func TestNetworkModeBridge(t *testing.T) {
...
+	defer agent.Cleanup()
+
+	params := &ecsapi.DescribeContainerInstancesInput{
+		Cluster:            &agent.Cluster,
+		ContainerInstances: []*string{&agent.ContainerInstanceArn},
+	}
+
+	resp, err := ECS.DescribeContainerInstances(params)
+	require.NoError(t, err)
+	require.NotEmpty(t, resp.ContainerInstances)
```
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: [multiple occurrences] can you also do a `require.Len(1, resp.ContainerInstances, ..)` here?

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 5]
```diff
@@ -618,3 +618,197 @@ func TestNetworkModeBridge(t *testing.T) {
...
+
+	attribMap := AttributesToMap(resp.ContainerInstances[0].Attributes)
+	assert.Equal(t, value, attribMap[key], "Values should match")
+}
+
+// TestCustomAttributesWithEmptyValue tests the ECS_INSTANCE_ATTRIBUTES
+// upon agent registration with key and empty value
+func TestCustomAttributesWithEmptyValue(t *testing.T) {
```
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: This is more of a unit/integ test for ECS; I'm not sure if it warrants a functional test in the Agent.

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 6]
```diff
@@ -618,3 +618,197 @@ func TestNetworkModeBridge(t *testing.T) {
...
+	require.NoError(t, err)
+	require.NotEmpty(t, resp.ContainerInstances)
+
+	attribMap := AttributesToMap(resp.ContainerInstances[0].Attributes)
+	assert.Equal(t, value, attribMap[key], "Values should match")
+}
+
+// TestCustomAttributesWithMultipleOptions tests the ECS_INSTANCE_ATTRIBUTES
+// upon agent registration with multiple key, value pairs
+func TestCustomAttributesWithMultipleOptions(t *testing.T) {
```
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: This is more of a unit/integ test for ECS; I'm not sure if it warrants a functional test in the Agent.

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 7]
```diff
@@ -618,3 +618,197 @@ func TestNetworkModeBridge(t *testing.T) {
...
+	for i := 0; i < numAttributes; i++ {
+		k := "key" + strconv.Itoa(i)
+		v := "val" + strconv.Itoa(i)
+		assert.Equal(t, v, attribMap[k], "Values should match")
+	}
+}
+
+// TestCustomAttributesWithMaxOptions tests the ECS_INSTANCE_ATTRIBUTES
+// upon agent registration with maximum number of supported key, value pairs
+func TestCustomAttributesWithMaxOptions(t *testing.T) {
```
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: This is more of a unit/integ test for ECS; I'm not sure if it warrants a functional test in the Agent.

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 8]
```diff
@@ -618,3 +618,197 @@ func TestNetworkModeBridge(t *testing.T) {
...
+	for i := 0; i < maxAttributes; i++ {
+		k := "key" + strconv.Itoa(i)
+		v := "val" + strconv.Itoa(i)
+		assert.Equal(t, v, attribMap[k], "Values should match")
+	}
+}
+
+// TestCustomAttributesWithMaxlenKey tests the ECS_INSTANCE_ATTRIBUTES
+// upon agent registration with a key of maximum supported length
+func TestCustomAttributesWithMaxlenKey(t *testing.T) {
```
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: This is more of a unit/integ test for ECS; I'm not sure if it warrants a functional test in the Agent.

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 9]
```diff
@@ -618,3 +618,197 @@ func TestNetworkModeBridge(t *testing.T) {
...
+	require.NoError(t, err)
+	require.NotEmpty(t, resp.ContainerInstances)
+
+	attribMap := AttributesToMap(resp.ContainerInstances[0].Attributes)
+	assert.Equal(t, value, attribMap[key], "Values should match")
+}
+
+// TestCustomAttributesWithMaxlenValue tests the ECS_INSTANCE_ATTRIBUTES
+// upon agent registration with a value of maximum supported length
+func TestCustomAttributesWithMaxlenValue(t *testing.T) {
```
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: This is more of a unit/integ test for ECS; I'm not sure if it warrants a functional test in the Agent.

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 10]
```diff
@@ -618,3 +618,197 @@ func TestNetworkModeBridge(t *testing.T) {
...
+	require.NoError(t, err)
+	require.NotEmpty(t, resp.ContainerInstances)
+
+	attribMap := AttributesToMap(resp.ContainerInstances[0].Attributes)
+	assert.Equal(t, value, attribMap[key], "Values should match")
+}
+
+// TestCustomAttributesWithMaxlen tests the ECS_INSTANCE_ATTRIBUTES
+// upon agent registration with a key, value of maximum supported length
+func TestCustomAttributesWithMaxlen(t *testing.T) {
```
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: This is more of a unit/integ test for ECS; I'm not sure if it warrants a functional test in the Agent.

File: agent/functional_tests/util/utils_unix.go [hunk 11]
```diff
@@ -150,6 +151,11 @@ func (agent *TestAgent) StartAgent() error {
 		Cmd: strings.Split(os.Getenv("ECS_FTEST_AGENT_ARGS"), " "),
 	}
 
+	// Append ECS_INSTANCE_ATTRIBUTES to dockerConfig
+	if attr := os.Getenv("ECS_INSTANCE_ATTRIBUTES"); attr != "" {
```
Comment (aaithal) [inline agent/functional_tests/util/utils_unix.go]: Since this is being used by the functional tests, can you name this `"ECS_FTEST_INSTANCE_ATTRIBUTES"` instead?

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 1 above)
Comment (vsiddharth) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: I personally did not want to have global constants and these lines here can be easily changed to incorporate changes to a specific feature improvement in the future. 

Also, now all the requirements for the test are local to a function.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 2 above)
Comment (vsiddharth) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Agreed

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 4 above)
Comment (vsiddharth) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: I can add them in but we only test the happy paths in our tests.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 3 above)
Comment (vsiddharth) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Agreed but as mentioned before, we only test the happy paths here!

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 5 above)
Comment (vsiddharth) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: It warrants a test as we can have scenarios with a valid key and no value.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 6 above)
Comment (vsiddharth) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: This test again is a genuine  use-case.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 7 above)
Comment (vsiddharth) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: This test again is a genuine functional use-case.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 8 above)
Comment (vsiddharth) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: This test again is a genuine functional use-case and tests the limits.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 9 above)
Comment (vsiddharth) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: This test again is a genuine functional use-case.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 10 above)
Comment (vsiddharth) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: This test again is a genuine functional use-case.

File: agent/functional_tests/util/utils_unix.go (hunk 11 above)
Comment (vsiddharth) [inline agent/functional_tests/util/utils_unix.go]: Agreed

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 1 above)
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: When we have those requirements, we can write this code if needed, I don't think we need it now as it's not really giving us any value.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 4 above)
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Happy path testing doesn't need to exclude all the validations that we can do.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 5 above)
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: It's an API correctness testing, which this package is not intended to do; the intention is to have coverage for the various APIs and fields that ECS supports.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 6 above)
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: It's an API correctness testing, which this package is not intended to do; the intention is to have coverage for the various APIs and fields that ECS supports.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 7 above)
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: It's an API correctness testing, which this package is not intended to do; the intention is to have coverage for the various APIs and fields that ECS supports.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 8 above)
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: It's an API correctness testing, which this package is not intended to do; the intention is to have coverage for the various APIs and fields that ECS supports.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 9 above)
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: It's an API correctness testing, which this package is not intended to do; the intention is to have coverage for the various APIs and fields that ECS supports.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 10 above)
Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: It's an API correctness testing, which this package is not intended to do; the intention is to have coverage for the various APIs and fields that ECS supports.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 1 above)
Comment (petderek) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: A problem with random strings is that it introduces potential to fail non-deterministically.  In practice, an engineer encountering that will likely re-run the tests instead of investigating the error. 
I really like the idea of trying to find unique edge cases, but I don't think functional tests are the right medium for that approach.

Comment (vsiddharth) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: @petderek In this case, the random string comes from a restricted and controlled set of characters which stick to the specs imposed. I was under the wrong impression that it makes sense for the functional tests to verify the boundary conditions under a genuine use-case. However, I've now removed all the edge case testing from here.

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 12]
```diff
@@ -618,3 +618,45 @@ func TestNetworkModeBridge(t *testing.T) {
 	err := networkModeTest(t, agent, "bridge")
 	require.NoError(t, err, "Networking mode bridge testing failed")
 }
+
+// TestCustomAttributesWithMaxOptions tests the ECS_INSTANCE_ATTRIBUTES
+// upon agent registration with maximum number of supported key, value pairs
+func TestCustomAttributesWithMaxOptions(t *testing.T) {
```
Comment (samuelkarp) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Is there any particular reason you did this only for Linux?  The same test should work on Windows as well (though you'll need to modify the Windows version of `StartAgent` separately).

Comment (aaithal) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: @vsiddharth It's fine if the answer here is that you're doing it in a separate PR, but we must look at adding a windows functional test for this as well.

Comment (vsiddharth) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: @samuelkarp I intend to do the windows test in a different PR

Comment (samuelkarp) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Are you following up immediately with Windows?  This will continue to be a gap in our coverage until we have a Windows test.

The test as written should work on Windows too.  The only thing you'd have to do is modify the Windows version of `StartAgent` to use the new `ECS_FTEST_INSTANCE_ATTRIBUTES` environment variable you specified.

**Comment by vsiddharth** [comment]
I've `re-run` all the tests manually on an ec2-instance and all the tests `succeed`.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 12 above)
Comment (vsiddharth) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: I will immediately follow up with a windows test and make the necessary changes

**Comment by liwenwu-amazon** [comment]
looks good to me!

File: agent/functional_tests/util/utils_unix.go [hunk 13]
```diff
@@ -238,3 +243,13 @@ func (agent *TestAgent) getBindMounts() []string {
 func (agent *TestAgent) Cleanup() {
 	agent.platformIndependentCleanup()
 }
+
+// AttributesToMap transforms a list of key, value attributes to return a map
+func AttributesToMap(attributes []*ecs.Attribute) map[string]string {
```
Comment (richardpen) [inline agent/functional_tests/util/utils_unix.go]: This is a general function not specific to linux platform, you should put in `utils.go`.

File: agent/functional_tests/util/utils_unix.go [hunk 14]
```diff
@@ -238,3 +243,13 @@ func (agent *TestAgent) getBindMounts() []string {
...
+	attributeMap := make(map[string]string)
+	attribs := attributes
```
Comment (richardpen) [inline agent/functional_tests/util/utils_unix.go]: I think you can just get rid of this line by using `for _, attribute := range attributes`

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 12 above)
Comment (richardpen) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: yeah, there will only be a small change to make it work on windows, can you do that in the same PR?

File: agent/functional_tests/util/utils_unix.go (hunk 11 above)
Comment (vsiddharth) [inline agent/functional_tests/util/utils_unix.go]: Reverted back to the old style to prevent code-duplication and run the test on both platforms

File: agent/functional_tests/util/utils_windows.go [hunk 15]
```diff
@@ -120,10 +120,14 @@ func (agent *TestAgent) StartAgent() error {
 	if TestDirectory := os.Getenv("ECS_WINDOWS_TEST_DIR"); TestDirectory != "" {
 		agentInvoke.Dir = TestDirectory
 	}
-	agentInvoke.Start()
+	err := agentInvoke.Start()
+	if err != nil {
+		agent.t.Logf("Agent start invocation failed")
```
Comment (aaithal) [inline agent/functional_tests/util/utils_windows.go]: nit: can you also print the error here?

File: agent/functional_tests/tests/functionaltests_test.go [hunk 16]
```diff
@@ -270,3 +275,45 @@ func networkModeTest(t *testing.T, agent *TestAgent, mode string) error {
...
+
+	resp, err := ECS.DescribeContainerInstances(params)
+	require.NoError(t, err)
+	require.NotEmpty(t, resp.ContainerInstances)
+	require.Len(t, resp.ContainerInstances, 1)
+
+	attribMap := AttributesToMap(resp.ContainerInstances[0].Attributes)
+	assert.NotEmpty(t, attribMap)
+
+	for i := 0; i < maxAttributes; i++ {
```
Comment (richardpen) [inline agent/functional_tests/tests/functionaltests_test.go]: I think it would be much better, you should check the variable you defined, but not the content.
```
for k, v := range customAttributes{
assert.Equal(t, v, attribMap[k], "Values should match")
}
```

Comment (richardpen) [inline agent/functional_tests/tests/functionaltests_test.go]: Just saw the `customAttributes` is not a map, ignore the comments.
